staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go

```diff
@@ -81,10 +81,7 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
 		if _, ok := versionedAttrs[invocation.Kind]; ok {
 			continue
 		}
```

Comment:
We're losing out on this optimization now... you could plumb the cached versionedAttrs map into `ShouldCallHook`, but that's getting messy. I'm not sure i see an elegant solution to this though

Reply:
I'll write a TODO for this but what if we split up the `ShouldCallWebhook` method? we have a seperate `CheckMatchConditions` method that can be called after `ShouldCallWebhook` or after rename maybe `CheckSelectors` and `CheckRules`. So right away  `CheckSelectors` and `CheckRules` run then after this check we either use the cached versionAttr or we construct one and we pass it to `CheckMatchConditions` ?

Reply:
Yeah, I think that's the way we need to do it. I'm also nervous about just leaving it as a TODO, since it will be a regression from the current approach even if the feature is disabled.

Reply:
> I'm also nervous about just leaving it as a TODO, since it will be a regression from the current approach even if the feature is disabled.

agreed... I don't think we can drop the existing optimization for a disabled alpha feature

passing an interface into ShouldCallHook that can produce versioned attrs for a given GVK is probably the smallest change for this PR (the impl could stuff things back into the map used here and the single item for mutating admission)